### PR TITLE
Store inline equations

### PR DESCRIPTION
**Pull request by oliverroick**
For inline equations, I have added a new node type in the frontend, which will be POSTed to the backend in the following form:

```
{
  "type": "equation-inline",
  "isInline": true,
  "children": [
    {"text": "e=mc^2}
  ]
}
```

With the changes in this PR and when I POST a document containing an inline equation, I get a success response but the inline equation is not returned with the document in the response. Likewise when I reload the page after saving the inline equation is not present. 

I'm sure I'm missing something, could you point me in the right direction.

### CODE REVIEW

**Comment by leothomas** [comment]
@oliverroick you were just missing a line including the `EquationInlineNode` type as a possible member of the `Section` type. 

I'm surprised that you were getting a success response from the POST, if you were using a node-type that wasn't explicitly allowed in the `Section` type. My guess is that Pydantic is ignoring the `type` value that it doesn't recognize and either: 
- Discarding the element all together
- Formatting it as an `Equation` type

I can look into that, but you should be good to go on this PR (for just inserting and retrieving the values from the DB). You still have to add handling for the inline equations to the documentation generation. 

I can look into that quickly

**Comment by leothomas** [comment]
It also look like Abstract and Plain Language Summary have to be handled as rich text in the PDF generation

File: app/schemas/document.py [hunk 1]
```diff
@@ -275,6 +275,7 @@ class SectionWrapper(BaseModel):
             TableBlockNode,
             SubsectionNode,
             EquationNode,
+            EquationInlineNode,
```
Comment (oliverroick) [inline app/schemas/document.py:278]: Is this correct, though? Should it be possible that an inline equation is a child of a section wrapper I was under the impression that only block-level nodes should be direct children of a section. 

Inline equations should be treated like other inline nodes, such as bold or italic, and should only be accepted as children of paragraphs. 

(It's also possible that I don't have a full understanding of how the nesting of these blocks work)

**Comment by oliverroick** [comment]
Thanks @leothomas , but this doesn't seem to work yet. 

Just to add a bit more detail, here's the full payload, I'm sending for the `abstract` field:

```json
{
  "abstract": {
    "children": [
      {
        "children": [
          {
            "text": " Lorem ipsum dolor sit amet, "
          },
          {
            "bold": true,
            "text": "consectetur"
          },
          {
            "text": " porttitor. "
          }
        ],
        "type": "p"
      },
      {
        "children": [
          {
            "text": "e=mc^2"
          }
        ],
        "isInline": false,
        "type": "equation"
      },
      {
        "children": [
          {
            "text": "Lorem ipsum dolor sit amet, consectetur "
          },
          {
            "children": [
              {
                "text": "latex~empty~equation"
              }
            ],
            "isInline": true,
            "type": "equation-inline"
          },
          {
            "text": " adipiscing elit. Vestibulum volutpat fermentum nisi feugiat mollis. Donec tristique sem vel felis semper tincidunt vel in sapien. Nunc ut tortor quis sem dictum imperdiet vitae sed tellus. Aliquam ultrices id risus vitae posuere. Nullam magna purus, tincidunt egestas dignissim et, varius consequat nisi. Donec eu leo mi. Duis sed nulla odio. Vivamus pharetra. "
          }
        ],
        "type": "p"
      }
    ]
  }
}
```

And the response I'm getting for the `abstract` field is:

```json
{
  "abstract": {
    "children": [
      {
        "type": "p",
        "children": [
          {
            "text": " Lorem ipsum dolor sit amet, "
          },
          {
            "text": "consectetur",
            "bold": true
          },
          {
            "text": " porttitor. "
          }
        ]
      },
      {
        "type": "equation",
        "children": [
          {
            "text": "e=mc^2"
          }
        ],
        "isInline": false
      },
      {
        "type": "p",
        "children": [
          {
            "text": "Lorem ipsum dolor sit amet, consectetur "
          },
          {
            "text": ""
          },
          {
            "text": " adipiscing elit. Vestibulum volutpat fermentum nisi feugiat mollis. Donec tristique sem vel felis semper tincidunt vel in sapien. Nunc ut tortor quis sem dictum imperdiet vitae sed tellus. Aliquam ultrices id risus vitae posuere. Nullam magna purus, tincidunt egestas dignissim et, varius consequat nisi. Donec eu leo mi. Duis sed nulla odio. Vivamus pharetra. "
          }
        ]
      }
    ]
  }
}
```

The inline is the second element of the third child. We're sending it as:

```json
{
  "children": [
    {
      "text": "latex~empty~equation"
    }
  ],
  "isInline": true,
  "type": "equation-inline"
}
```

But it comes back as:

```json
{
  "text": ""
}
```

Is it possible that I need to structure inline elements in a different way? For example, does the `text` field need to sit at the same level as the `type` key?

File: app/schemas/document.py (hunk 1 above)
Comment (leothomas) [reply app/schemas/document.py]: You're absolutely correct, thanks for catching that!

**Comment by leothomas** [comment]
@oliverroick thanks for all the clarifications. Just added a few more edits, you should be good to go now!

Note - I removed the boolean `isInLine` field since the `type` field already denotes between `equation` and `equation-inline`. 

The reason why the inline equations were showing up as `{"text":""}` is because of a peculiar behaviour in Pydantic (described [here](https://developmentseed.slack.com/archives/C04KHH66Q/p1620322286123200)): the `div` type's `children` field is defined as `List[Union[linkNode, refNode, textLeaf, equationInline]]` - so pydantic will build the object that _first_ matches the input rather than the object that _best_ matches the input (in this case `textLeaf` rather than `equationInline`). The fix is to switch the order of the possible child nodes and let `textLeaf` act as a catch for all the inputs that don't already match to something else. 

The payload should look like: 
```json
"document": {
    "abstract": {
        "children": [
            {
                "type": "p",
                "children": [
                    {
                        "text": "A brief and interesting overview of the themes presented in this document are as follows: At vero eos et accusamus et iusto odio dignissimos ducimus qui blanditiis praesentium voluptatum deleniti atque corrupti quos dolores et quas molestias excepturi sint occaecati cupiditate non provident, similique sunt in culpa qui officia deserunt mollitia animi, id est laborum et dolorum fuga."
                    },
                    {
                        "type": "equation-inline",
                        "children": [
                            {
                                "text": "\\delta = \\sum_{n=1}^{\\infty} 2^{-n} = 1"
                            }
                        ],
                    },
                ],
            }
        ]
    }
}
```

Screen shots of generated documents: 

regular: 
image

journal: 
image

**Comment by oliverroick** [comment]
@leothomas I've adapted the front-end accordingly and this works fine now. Can we get this approved?

**Pull request merged**